src/WorkerChannel.ts

```diff
@@ -33,32 +29,32 @@ export class WorkerChannel {
         });
     }
 
-    /**
-     * Register a patching function to be run before User Function is executed.
-     * Hook should return a patched version of User Function.
-     */
-    public registerBeforeInvocationRequest(beforeCb: InvocationRequestBefore): void {
-        this._invocationRequestBefore.push(beforeCb);
-    }
-
-    /**
-     * Register a function to be run after User Function resolves.
-     */
-    public registerAfterInvocationRequest(afterCb: InvocationRequestAfter): void {
-        this._invocationRequestAfter.push(afterCb);
+    public registerHook(hookName: string, callback: HookCallback): Disposable {
+        const hooks = this.getHooks(hookName);
+        hooks.push(callback);
+        return new Disposable(() => {
+            const index = hooks.indexOf(callback);
+            if (index > -1) {
+                hooks.splice(index, 1);
+            }
+        });
     }
 
-    public runInvocationRequestBefore(context: Context, userFunction: Function): Function {
-        let wrappedFunction = userFunction;
-        for (const before of this._invocationRequestBefore) {
-            wrappedFunction = before(context, wrappedFunction);
+    public async executeHooks(hookName: string, context: {}): Promise<void> {
+        const callbacks = this.getHooks(hookName);
+        for (const callback of callbacks) {
+            await callback(context);
         }
-        return wrappedFunction;
     }
 
-    public runInvocationRequestAfter(context: Context) {
-        for (const after of this._invocationRequestAfter) {
-            after(context);
+    private getHooks(hookName: string): HookCallback[] {
+        switch (hookName) {
+            case 'preInvocation':
+                return this._preInvocationHooks;
+            case 'postInvocation':

```

Comment:
## enums

TypeScript has great support for [string literal types](https://www.typescriptlang.org/docs/handbook/literal-types.html) (`'a' | 'b'`) and I frankly never use enums. At least in TypeScript, enums are super clunky and can have weird behavior if the first value of the enum is `0` because users so often do `if (val) {`. I recommend you try out string literal types (if you haven't already) and see if you can cause a bug without TypeScript throwing an error first. But yes `string` by itself is not type safe and I agree with that

Reply:
Yeah totally agree with this. That's what I meant with a "custom enum-style type," similar to what we use for `HttpMethod` for example. I just didn't know the technical term was string literal types